event-web-framework/draft.markdown

```diff
@@ -0,0 +1,941 @@
+
+```lisp
+	(defclass request ()
+	  ((resource :accessor resource :initarg :resource)
+	   (headers :accessor headers :initarg :headers :initform nil)
+	   (parameters :accessor parameters :initarg :parameters :initform nil)))
+```
+
+We've now seen how our clients can send requests and have them interpreted and handled by our server. The last thing we have to implement as part of our core server interface is the capability to write responses back to the client.
+
+### Rendering Responses
+
+Before we discuss rendering responses, we have to consider that there are two kinds of responses that we may be returning to our clients. The first is a "normal" HTTP response, complete with HTTP headers and body. We represent these kinds of responses with instances of the `response` class:
+
+```lisp
+(defclass response ()
+  ((content-type
+    :accessor content-type :initform "text/html" :initarg :content-type)
+   (charset
+    :accessor charset :initform "utf-8")
+   (response-code
+    :accessor response-code :initform "200 OK" :initarg :response-code)
+   (keep-alive?
+    :accessor keep-alive? :initform nil :initarg :keep-alive?)
+   (body
+    :accessor body :initform nil :initarg :body)))
+```
+
+The second is an [SSE message](http://www.w3.org/TR/eventsource/), which we will use to send an incremental update to our clients.
+
+```lisp
+(defclass sse ()
+  ((id :reader id :initarg :id :initform nil)
+   (event :reader event :initarg :event :initform nil)
+   (retry :reader retry :initarg :retry :initform nil)
+   (data :reader data :initarg :data)))
+```
+
+We'll send an HTTP response whenever we receive a full HTTP request; however, how do we know when and where to send SSE messages without an originating client request?
+
+A simple solution is to register _channels_[^defparameter], to which we'll subscribe `socket`s as necessary.
+
+```lisp
+(defparameter *channels* (make-hash-table))
+
+(defmethod subscribe! ((channel symbol) (sock usocket))
+  (push sock (gethash channel *channels*))
+  nil)
+```
+
+[^defparameter]: We're incidentally introducing some new syntax here. This is our way of declaring a mutable variable. It has the form `(defparameter <name> <value> <optional docstring>)`.
+
+We can then `publish!` notifications to said channels as soon as they become available.
+
+```lisp
+(defmethod publish! ((channel symbol) (message string))
+  (awhen (gethash channel *channels*)
+	 (setf (gethash channel *channels*)
+	       (loop with msg = (make-instance 'sse :data message)
+		  for sock in it
+		  when (ignore-errors
+			 (write! msg sock)
+			 (force-output (socket-stream sock))
+			 sock)
+		  collect it))))
+```
+
+In `publish!`, we call `write!` to actually write an `sse` to a socket. We'll also need a specialization of `write!` on `response`s to write full HTTP responses as well. Let's handle the HTTP case first.
+
+```lisp
+(defmethod write! ((res response) (socket usocket))
+  (handler-case
+      (with-timeout (.2)
```

Comment:
Implementation dependant. You can see the details at https://github.com/gwkkwg/trivial-timeout/blob/master/dev/with-timeout.lisp#L16-L120

(Some implementations provide a primitive timeout operation, which seems to be used where available. Others don't, and in those, it seems to spawn either a thread or a process.)

I'm not a hundred percent sure this is something we need to worry about in the first place; Common Lisp `write-sequence` doesn't seem to block on a foreign socket, even if data isn't read out of it. Maybe we just note that instead of including a separate external library?)